backend/local/backend_local.go

```diff
@@ -245,6 +258,159 @@ func (b *Local) contextFromPlanFile(pf *planfile.Reader, opts terraform.ContextO
 	return tfCtx, snap, diags
 }
 
+// interactiveCollectVariables attempts to complete the given existing
+// map of variables by interactively prompting for any variables that are
+// declared as required but not yet present.
+//
+// If interactive input is disabled for this backend instance then this is
+// a no-op. If input is enabled but fails for some reason, the resulting
+// map will be incomplete. For these reasons, the caller must still validate
+// that the result is complete and valid.
+//
+// This function does not modify the map given in "existing", but may return
+// it unchanged if no modifications are required. If modifications are required,
+// the result is a new map with all of the elements from "existing" plus
+// additional elements as appropriate.
+//
+// Interactive prompting is a "best effort" thing for first-time user UX and
+// not something we expect folks to be relying on for routine use. Terraform
+// is primarily a non-interactive tool and so we prefer to report in error
+// messages that variables are not set rather than reporting that input failed:
+// the primary resolution to missing variables is to provide them by some other
+// means.
+func (b *Local) interactiveCollectVariables(ctx context.Context, existing map[string]backend.UnparsedVariableValue, vcs map[string]*configs.Variable, uiInput terraform.UIInput) map[string]backend.UnparsedVariableValue {
+	var needed []string
+	if b.OpInput && uiInput != nil {
+		for name, vc := range vcs {
+			if !vc.Required() {
+				continue // We only prompt for required variables
+			}
+			if _, exists := existing[name]; !exists {
+				needed = append(needed, name)
+			}
+		}
+	} else {
+		log.Print("[DEBUG] backend/local: Skipping interactive prompts for variables because input is disabled")
+	}
+	if len(needed) == 0 {
+		return existing
+	}
+
+	log.Printf("[DEBUG] backend/local: will prompt for input of unset required variables %s", needed)
+
+	// If we get here then we're planning to prompt for at least one additional
+	// variable's value.
+	sort.Strings(needed) // prompt in lexical order
```

Comment:
This is a question, not a suggestion, despite the phrasing: 

If `needed` wasn't sorted, what order would the variables show up in? If the variables are (at least per-fil) in the same order that they are in the configuration files it might make more sense to the user if the variables are presented in that order. I'm thinking of scenarios where one has a lot of related variables lumped together in each file and where the variables make more sense when thought of as a whole (as opposed to seeing them "split up" in lexical order).

Reply:
The variable definitions and values both live in maps so the _default_ sort here would be pseudorandom.

In principle we could try to sort them by their source locations in order to recover the order in source code that you're thinking about here. So far the only precedent we've had for sorting things in lexical order is that we sort diagnostic messages that way in the hope of them appearing in an order that allows you to work through them one by one; I expect we could adapt similar logic here but we'd probably need to generalize some logic for actually sorting by the source locations so that it can be used from multiple locations.

I'm open to doing that, but I'd rather pursue it as a separate change, since my goal here was to retain the same user-facing behavior we had before and just move the logic to a different place in the code.